zsh-window-title.zsh

```diff
@@ -4,139 +4,182 @@
+
+    # update window title when changing directories
+    add-zsh-hook chpwd __zsh-window-title:chpwd-update  # New hook for updating title on directory change
 }
 
 __zsh-window-title:init() {
-	'builtin' 'emulate' -LR zsh
+    'builtin' 'emulate' -LR zsh
 
-	'builtin' 'typeset' -gi ZWT_DEBUG=${ZWT_DEBUG:-$__zwt_debug_default}
+    'builtin' 'typeset' -gi ZWT_DEBUG=${ZWT_DEBUG:-$__zwt_debug_default}
 
-	'builtin' 'typeset' -gi ZSH_WINDOW_TITLE_DEBUG=${ZSH_WINDOW_TITLE_DEBUG:-$__zsh_window_title_debug_default}
+    'builtin' 'typeset' -gi ZSH_WINDOW_TITLE_DEBUG=${ZSH_WINDOW_TITLE_DEBUG:-$__zsh_window_title_debug_default}
 
-	__zsh-window-title:debugger
+    __zsh-window-title:debugger
 
-	'builtin' 'typeset' -gi ZSH_WINDOW_TITLE_DIRECTORY_DEPTH=${ZSH_WINDOW_TITLE_DIRECTORY_DEPTH:-$__zsh_window_title_directory_depth_default}
+    'builtin' 'typeset' -gi ZSH_WINDOW_TITLE_DIRECTORY_DEPTH=${ZSH_WINDOW_TITLE_DIRECTORY_DEPTH:-$__zsh_window_title_directory_depth_default}
 
-	__zsh-window-title:precmd
+    __zsh-window-title:precmd
 
-	'builtin' 'autoload' -U add-zsh-hook
+    'builtin' 'autoload' -U add-zsh-hook
 
-	__zsh-window-title:add-hooks
+    __zsh-window-title:add-hooks
 }
 
 __zsh-window-title:precmd() {
-	'builtin' 'emulate' -LR zsh
-	__zsh-window-title:debugger
-
-	local title=$(print -P "%$ZSH_WINDOW_TITLE_DIRECTORY_DEPTH~")
+    'builtin' 'emulate' -LR zsh
+    __zsh-window-title:debugger
 
-	'builtin' 'echo' -ne "\033]0;$title\007"
+    local current_dir=$(print -P "%$ZSH_WINDOW_TITLE_DIRECTORY_DEPTH~")  # Truncated current directory
+    __zsh-window-title:update "$current_dir"
 }
 
 __zsh-window-title:preexec() {
-	'builtin' 'emulate' -LR zsh
-	__zsh-window-title:debugger
+    'builtin' 'emulate' -LR zsh
+    __zsh-window-title:debugger
+
+    local cmd_first_part=${1[(w)1]}  # Store the first word of the command
+    local current_dir=$(print -P "%$ZSH_WINDOW_TITLE_DIRECTORY_DEPTH~")  # Truncated current directory
+    __zsh-window-title:update "$current_dir" "$cmd_first_part"
+}
+
+__zsh-window-title:update() {
+    local title_content="$1"
+    if [[ -n "$2" ]]; then
+        title_content="$title_content $2"
+    fi
+
+    # Set terminal title (xterm compatible)
+    printf "\033]0;%s\007" "$title_content"
+    printf "\033]0;%s\033\\" "$title_content"
 
-	local title=$(print -P "%$ZSH_WINDOW_TITLE_DIRECTORY_DEPTH~ - ${1[(w)1]}")
+    # Set tmux status bar title (xterm compatible)
+    printf "\033]2;%s\007" "$title_content"
+    printf "\033]2;%s\033\\" "$title_content"
+
+    # Set screen window name (specific to GNU Screen)
+    printf "\033k%s\033\007" "$title_content"
+    printf "\033k%s\033\\" "$title_content"
+
+    # Rename tmux window
+    if [ -n "$TMUX" ]; then
+        tmux rename-window "$title_content"
+    fi
+}
+
+
+__zsh-window-title:chpwd-update() {
+    local current_dir=$(print -P "%15<..<%~")  # Truncated current directory
```

Comment:
What's the motivation for not using the existing `ZSH_WINDOW_TITLE_DIRECTORY_DEPTH` pattern here?

Same question for __zsh-window-title:preexec-update and for __zsh-window-title:precmd-update

Reply:
To limit the length of the directory, so that it doesnt display more than 15 characters long